File: ruby/count_primes.rb
# Description:
#
# Count the number of prime numbers less than a non-negative number, n.

# @param {Integer} n
# @return {Integer}
def is_prime?(n)
  return false if n <= 1
  return true if n == 2

  (2..n / 2).each do |divisor|
    return false if n % divisor == 0
  end

  true
end

def count_primes(n)
  (2...n).select { |n| is_prime?(n) }.count
end


p is_prime?(2) == true
p is_prime?(3) == true
p is_prime?(4) == false
p is_prime?(5) == true
p is_prime?(8) == false
p is_prime?(7) == true

p count_primes(4)

File: javascript/countPrimes.js
// Count the number of prime numbers less than a non-negative number, n.
/**
 * @param {number} n
 * @return {number}
 */
var countPrimes = function(n) {
  var primeCache = {};
  for (var i = 0; i < n; i++) {
    primeCache[i] = true;
  }

  for (var i = 2; i * i < n; i++) {
    for (var j = i * i; j < n; j += i) {
      primeCache[j] = false;
    }
  }

  var count = 0;
  for (var k = 2; k < n; k++) {
    if (primeCache[k]) count++;
  }

  return count;
};


console.log(countPrimes(3) === 1);
console.log(countPrimes(5) === 2);
console.log(countPrimes(6) === 3);
